src/pool-bin/libraries/BinHooks.sol

```diff
@@ -0,0 +1,157 @@
+        ) {
+            revert Hooks.HookPermissionsValidationError();
+        }
+
+        if (
+            key.parameters.hasOffsetEnabled(HOOKS_AFTER_SWAP_RETURNS_DELTA_OFFSET)
+                && !key.parameters.hasOffsetEnabled(HOOKS_AFTER_SWAP_OFFSET)
+        ) {
+            revert Hooks.HookPermissionsValidationError();
+        }
+
+        if (
+            key.parameters.hasOffsetEnabled(HOOKS_AFTER_MINT_RETURNS_DELTA_OFFSET)
+                && !key.parameters.hasOffsetEnabled(HOOKS_AFTER_MINT_OFFSET)
+        ) {
+            revert Hooks.HookPermissionsValidationError();
+        }
+
+        if (
+            key.parameters.hasOffsetEnabled(HOOKS_AFTER_BURN_RETURNS_DELTA_OFFSET)
+                && !key.parameters.hasOffsetEnabled(HOOKS_AFTER_BURN_OFFSET)
+        ) {
+            revert Hooks.HookPermissionsValidationError();
+        }
+    }
+
+    function afterMint(
+        PoolKey memory key,
+        IBinPoolManager.MintParams memory params,
+        BalanceDelta delta,
+        bytes calldata hookData
+    ) internal returns (BalanceDelta callerDelta, BalanceDelta hookDelta) {
+        IBinHooks hooks = IBinHooks(address(key.hooks));
+        callerDelta = delta;
+
+        if (key.parameters.shouldCall(HOOKS_AFTER_MINT_OFFSET, hooks)) {
+            bytes4 selector;
+            (selector, hookDelta) = hooks.afterMint(msg.sender, key, params, delta, hookData);
+
+            if (selector != IBinHooks.afterMint.selector) {
+                revert Hooks.InvalidHookResponse();
+            }
+
+            if (
+                key.parameters.hasOffsetEnabled(HOOKS_AFTER_MINT_RETURNS_DELTA_OFFSET)
+                    && hookDelta != BalanceDeltaLibrary.ZERO_DELTA
+            ) {
+                callerDelta = callerDelta - hookDelta;
+            }
+        }
+    }
+
+    function afterBurn(
+        PoolKey memory key,
+        IBinPoolManager.BurnParams memory params,
+        BalanceDelta delta,
+        bytes calldata hookData
+    ) internal returns (BalanceDelta callerDelta, BalanceDelta hookDelta) {
+        IBinHooks hooks = IBinHooks(address(key.hooks));
+        callerDelta = delta;
+
+        if (key.parameters.shouldCall(HOOKS_AFTER_BURN_OFFSET, hooks)) {
+            bytes4 selector;
+            (selector, hookDelta) = hooks.afterBurn(msg.sender, key, params, delta, hookData);
+
+            if (selector != IBinHooks.afterBurn.selector) {
+                revert Hooks.InvalidHookResponse();
+            }
+
+            if (
+                key.parameters.hasOffsetEnabled(HOOKS_AFTER_BURN_RETURNS_DELTA_OFFSET)
+                    && hookDelta != BalanceDeltaLibrary.ZERO_DELTA
+            ) {
+                callerDelta = callerDelta - hookDelta;
+            }
+        }
+    }
+
+    function beforeSwap(PoolKey memory key, bool swapForY, uint128 amountIn, bytes calldata hookData)
+        internal
+        returns (uint128 amountToSwap, int128 hookDeltaSpecified)
```

Comment:
This is `uint128` because unlike cl-pool, in bin-pool users are always specifying exactlyIn amount